c10/core/TensorTypeSet.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <c10/core/TensorTypeId.h>
+#include <c10/util/llvmMathExtras.h>
+#include <c10/util/Exception.h>
+#include <ostream>
+
+namespace c10 {
+
+// A representation of a set of TensorTypeIds.  A tensor may
+// have multiple tensor type ids, e.g., a Variable tensor
+// can also be a CPU tensor; the TensorTypeSet specifies what
+// type ids apply.  The internal representation is as a 64-bit
+// bit set (this means only 64 tensor type ids are supported).
+//
+// At the moment, there are no nontrivial uses of this set; tensors
+// are always singletons.  In the near future, this set will represent
+// variable? + tensor type id.  In the far future, it will be
+// variable? + profiling? + tracing? + lazy? + tensor type id.
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
+public:
+  TensorTypeSet() {}
+  explicit TensorTypeSet(TensorTypeId t)
+    : repr_(t == TensorTypeId::UndefinedTensorId
+              ? 0
+              : 1ULL << (static_cast<uint8_t>(t) - 1)) {}
+  // Test if a TensorTypeId is in the set
+  bool has(TensorTypeId t) const {
+    TORCH_INTERNAL_ASSERT(t != TensorTypeId::UndefinedTensorId);
+    return static_cast<bool>(repr_ & TensorTypeSet(t).repr_);
+  }
+  // Perform set union
+  TensorTypeSet operator|(TensorTypeSet other) const {
+    return TensorTypeSet(repr_ | other.repr_);
+  }
+  // Perform set equality
+  bool operator==(TensorTypeSet other) const {
+    return repr_ == other.repr_;
+  }
+  // Add a TensorTypeId to the TensorTypeId set.  Does NOT mutate,
+  // returns the extended TensorTypeSet!
+  C10_NODISCARD TensorTypeSet add(TensorTypeId t) const {
+    return *this | TensorTypeSet(t);
+  }
+  // Remove a TensorTypeId from the TensorTypeId set.  This is
+  // generally not an operation you should be doing (it's
+  // used to implement operator<<)
+  C10_NODISCARD TensorTypeSet remove(TensorTypeId t) const {
+    return TensorTypeSet(repr_ & ~TensorTypeSet(t).repr_);
+  }
+  // Is the set empty?  (AKA undefined tensor)
+  bool empty() const {
```

Comment:
I'm not sure I totally understand where we still use undefined tensors, but it feels a little strange to provide this as a BitSet interface when there is this outlier with semantic meaning.

Reply:
This is not explained clearly in the patch (yet!), but here is how I think of it.

What we call `TensorTypeId` is actually `optional<RealTensorTypeId>`. The `nullopt` case corresponds to `UndefinedTensorId`. `firstTypeId` returns an `optional<RealTensorTypeId>` because there may not be any RealTensorTypeIds in the set: so you have to return a `nullopt` (aka `UndefinedTensorId` in that case). So, the mapping is:

* `UndefinedTensorId ==> empty TensorTypeSet`
* `CPUTensorId ==> singleton TensorTypeSet containing only CPUTensorId`

`UndefinedTensor` never actually is present in the bitset.

Reply:
you don't have to do this, but it seems nice to do an inventory of places where we actually use UndefinedTensorId.  I bet we can rid of most/all of them by using optional and then we wouldn't have to do this.